### PR TITLE
[Prototype][MRV2] Stage DeepSeek decode tokens for imbalanced DBO graph replay

### PR DESCRIPTION
**Pull request by algebra-MCX**
## Purpose

Related issue: #57177; RFC: #50738. Built on MRV2 FULL-graph DBO support in #51700, with branch base `73d2a8cf86c46e21af723ffd34ce0d56668fdd14`.

During imbalanced DP decode, a rank with at most half the captured token capacity leaves its second physical microbatch empty. This prototype stages real tokens into both captured halves to retain FULL replay. For 63 real tokens in a 128-token graph, 32 rows stay at offset 0 and 31 move to offset 64.

After replay, restore hidden states, input IDs, positions, block tables, slot mappings and padding flags to logical request order before sampling. Restoring hidden states alone leaves the sampler reading padding IDs and positions. When returning routed experts, write each microbatch into its own region of the capture buffer and compact routing records before taking the existing private snapshot, avoiding overwritten or misaligned expert records.

### Scope

- DeepSeek-V2, FLASH_ATTN_MLA, NIXL EP, TP1/PP1, two microbatches and single-token decode. Real DP2 and DP4 execution has been validated with DeepSeek-V2-Lite BF16.
- Context parallelism, LoRA, speculation, EPLB, KV transfer and prompt embeddings are excluded from conditional staging. Eligible even graph capacities are supported, including the equality boundary; unsupported layouts retain eager fallback.
- Captured buffer addresses remain stable. Returning routed experts retains DBO within this supported scope; outside it, that combination disables DBO with a warning.
- This remains a restricted prototype for design review. General model/backend support, cross-layout numerical equivalence and universal performance gains are not claimed.

### Related work

Open-PR searches for `57177 in:body` and `DBO staging` were refreshed on 2026-09-20. This updates the existing #57184 rather than opening another implementation. Open #57020 handles the empty-tail equality boundary by falling back to eager; this prototype additionally stages eligible real tokens to avoid that fallback. The overlapping predicate must be reconciled before merge. Other search matches concern DeepSeek-V4.1/Engram, NCCL reduce-scatter or expert offloading.

## Test Plan

Using the configured vLLM virtual environment, run from the PR worktree:

```bash
.venv/bin/python -m pytest \
  tests/v1/worker/test_gpu_ubatch_slicing.py \
  tests/model_executor/test_routed_experts_capture.py \
  tests/v1/worker/test_gpu_model_runner_v2.py \
  tests/v1/worker/test_gpu_model_runner_v2_cudagraph_profiling.py -q

.venv/bin/python -m pytest \
  tests/kernels/moe/test_routed_experts_capture_monolithic.py -q

.venv/bin/pre-commit run --files \
  vllm/v1/worker/gpu/dp_utils.py \
  vllm/v1/worker/gpu/model_runner.py \
  vllm/v1/worker/gpu/ubatch_utils.py \
  vllm/model_executor/layers/fused_moe/routed_experts_capturer.py \
  tests/v1/worker/test_gpu_ubatch_slicing.py \
  tests/model_executor/test_routed_experts_capture.py \
  tests/kernels/moe/test_routed_experts_capture_monolithic.py
git diff --check
```

The recorded workspace used the shared `../vllm-mrv2-dbo/.venv` and explicitly selected this PR worktree on `PYTHONPATH`.

### Regression tests and source provenance

- H200 regression suite: **105 passed**, including actual CUDA graph replay, capture-buffer address stability, writes across DBO yields, rank swaps, 63/64/65 boundaries, fallback paths and the production execute/sample input contract. The final graph-test rerun passed both cases. The monolithic module was skipped because it requires SM100.
- Before the routing fix, the targeted new tests produced **8 failed, 6 passed**. The input-restoration regression independently produced **3 failed, 1 passed** without restoration; the 65-token non-staging control passed.
- Changed-file pre-commit checks passed, including Ruff and mypy. The sandbox CPU rerun of slicing and capturer tests produced **66 passed, 19 skipped**; this is separate from the recorded H200 run.
- The complete six-file local fix patch matched the validated artifact byte-for-byte (SHA256 `f6d19ffdc25d58092c49bb29e5cec658efc551cec51cc245cb0478481a0f57c6`). Raw serving records were revalidated on 2026-09-20, including AST equality between saved production sources and the submitted source.

### Real-model serving validation

DeepSeek-V2-Lite revision `604d5664dddd88a0433dbae533b7fe9472482de0`, H200, BF16, TP1/PP1, NIXL EP, FLASH_ATTN_MLA, batched_triton, MRV2 and FULL CUDA graphs. Requests used temperature 0.8, top_p 0.95, top_k 40, explicit seeds, ignore_eos=False and min_tokens 0/16. Loads included DP2 `63/97`, `97/63`, `64/97` and DP4 `63/97/80/80`, `97/63/80/80`, `32/32/32/224`.

| Run | Successful requests | Sampling observations / staged | FULL/Eager observations / staged | Expert-ID comparisons |
| --- | ---: | ---: | ---: | ---: |
| DP2, DBO without routing return | 1,282 | 246 / 78 | 96 / 36 | — |
| DP2, DBO with routing return | 1,282 | 235 / 77 | 96 / 36 | 96 |
| DP4, DBO with routing return | 2,560 | 489 / 107 | 188 / 54 | 188 |

All **5,124 requests completed**. Observed sampling IDs, positions, sampled tokens, min-token predicates and processing masks had zero mismatches. Same-physical-layout FULL/Eager hidden states matched exactly and complete KV snapshots were byte-identical. Routing slot checks passed; independent router top-k references matched all 26 MoE layers in the observed comparisons. All 3,842 routing-enabled HTTP responses passed shape/range checks, covering 856,024 token routing records. These are sampled numerical comparisons plus response integrity checks, not an independent numerical reference for every returned token.

An earlier input-restoration revision completed a 32-question GSM8K evaluation: 6/32 correct, invalid_rate=0, temperature 0.8, seed 57184, 0-shot, max_tokens 512. It has no matched quality baseline and does not establish quality parity for this revision.

### Performance and remaining limits

Current repaired-source DP2 timing on the same otherwise idle H200 pair, three seeds, budget 8192, directed `63/97` traffic and no diagnostic probes: DBO without routing return **1438.73 ± 27.56 tok/s**; DBO with routing return **1415.68 ± 22.83 tok/s**; DBO disabled with routing return **2481.46 ± 10.63 tok/s** (means ± sample standard deviations). Thus the combined feature was 1.60% below the DBO path without routing return and 42.95% below the no-DBO control. This is an end-to-end feature comparison, not an isolated compaction cost or a speedup claim for staging. A separate budget-256 experiment on shared GPUs also observed DBO slower; external load prevents a clean budget-to-budget comparison or precise attribution. Further measured profiling has not yet established the critical-path cause.

Same-layout FULL/Eager checks do not resolve the previously observed cross-layout BF16 shared-expert numerical differences. Services exited successfully; known UCX and API output-handler diagnostics remained during shutdown, with no serving-time traceback in the accepted runs. No claim of warning-free shutdown is made.

Raw source snapshots, commands, responses, worker observations and logs are retained locally; there is no public raw-artifact URL.

## AI assistance and human review

OpenAI Codex assisted with implementation, tests, experiments, review and PR preparation. The submitting user reviewed the reported changes and validation and explicitly requested this update. Human line-by-line review and human execution of the relevant tests have not been independently confirmed; the repository's human-review requirements remain applicable before merge.

### CODE REVIEW

File: vllm/v1/worker/gpu/ubatch_utils.py [hunk 1]
```diff
@@ -301,6 +343,26 @@ def __init__(
         self.model_state = model_state
         self.attn_groups = attn_groups
         self.kv_cache_config = kv_cache_config
+        # Experimental scope: DeepSeek-V2 with FlashAttention MLA and NIXL only.
```
Comment (specture724) [inline vllm/v1/worker/gpu/ubatch_utils.py]: As you mentioned, this code is only a prototype for DeepSeek V2 under very spcific configuration. I'd like to make it more general before merging.

File: vllm/v1/worker/gpu/model_runner.py [hunk 2]
```diff
@@ -1899,6 +1896,13 @@ def execute_model(
                 **connector_kwargs, attn_metadata=attn_metadata
             )
             model_output = self.cudagraph_manager.run_fullgraph(batch_desc)
+            if ubatch_state is not None and ubatch_state.real_split is not None:
```
Comment (specture724) [inline vllm/v1/worker/gpu/model_runner.py]:
```python
if ubatch_state is not None and ubatch_state.staged_rows is not None:
        staged_rows = ubatch_state.staged_rows
        model_output[: staged_rows.numel()] = model_output[staged_rows]
```
and just delete `compact_conditional_output`.

Edit: Input buffers are rewritten in place. Anything that reads them after the forward pass and expects the original order would get wrong results instead of falling back to eager. Routed-experts capture is one example. The routed-experts capturer (`model_runner.py:1964-1966`) reads the staged slot_mappings, where rows `[first:n]` are now -1. Its outputs are never put back in order, and the eligibility check doesn't exclude this feature.

File: vllm/v1/worker/gpu/dp_utils.py [hunk 3]
```diff
@@ -127,10 +127,24 @@ def sync_cudagraph_and_dp_padding(
...
                 )
-                if 2 * int(num_tokens_across_dp.min()) < ubatch_desc.num_tokens:
-                    # If one rank has an empty second microbatch, run without
-                    # CUDA graphs.
-                    ubatch_desc = None
+                if 2 * int(num_tokens_across_dp.min()) <= ubatch_desc.num_tokens:
+                    # Only the restricted decode prototype can stage real rows
+                    # into both halves of the already captured static layout.
+                    runner = cudagraph_manager.ubatch_runner
+                    can_stage = (
```
Comment (specture724) [inline vllm/v1/worker/gpu/dp_utils.py:134]: `can_stage` has several redundant conditions

Five of the ten conditions in `dp_utils.py` are implied by others:

- `num_ubatches == 2` — already part of `conditional_real_split`.
- `torch.all(num_reqs_across_dp == num_tokens_across_dp)` — implied by `synced_uniform_token_count == 1` (every rank reported a uniform query length of 1).
- `ubatch_desc.num_reqs == ubatch_desc.num_tokens` — a FULL microbatched descriptor only comes from `_maybe_ubatch_twin`, whose `uniform_token_count = num_tokens //
num_reqs` must match the synced count of 1.
- `ubatch_desc.num_tokens >= 4` — implied by `min >= 2` and `2 * min <= num_tokens`.
- `runner is not None` + `getattr(...)` — a FULL microbatched descriptor only exists when `ubatch_runner` is set; `runner` is typed, so `getattr` is only there for the `SimpleNamespace` in the test.

The same eligibility rules are also repeated in `prepare()` and in the asserts of `stage_conditional_decode`. `prepare()` doesn't need its own copy: if the DP sync picked FULL and this rank's last microbatch is empty, the DP-level staging check must already have passed.

File: vllm/v1/worker/gpu/ubatch_utils.py [hunk 4]
```diff
@@ -44,6 +44,48 @@ class UBatchState(NamedTuple):
...
+    slot_mappings[:, first:capacity].fill_(-1)
+    slot_mappings[:, capacity + second : 2 * capacity].fill_(-1)
+    input_batch.is_padding[:first].fill_(False)
+    input_batch.is_padding[first:capacity].fill_(True)
+    input_batch.is_padding[capacity : capacity + second].fill_(False)
+    input_batch.is_padding[capacity + second : 2 * capacity].fill_(True)
+    return first, second
+
+
+def compact_conditional_output(
```
Comment (specture724) [inline vllm/v1/worker/gpu/ubatch_utils.py]: ditto

File: vllm/v1/worker/gpu/ubatch_utils.py [hunk 5]
```diff
@@ -301,6 +343,26 @@ def __init__(
         self.model_state = model_state
         self.attn_groups = attn_groups
         self.kv_cache_config = kv_cache_config
+        # Experimental scope: DeepSeek-V2 with FlashAttention MLA and NIXL only.
+        self.conditional_real_split = (
+            vllm_config.model_config.hf_config.model_type == "deepseek_v2"
```
Comment (specture724) [inline vllm/v1/worker/gpu/ubatch_utils.py]: As you mentioned, this code is only a prototype for DeepSeek V2 under very spcific configuration. I'd like to make it more general before merging.

File: vllm/v1/worker/gpu/ubatch_utils.py [hunk 6]
```diff
@@ -301,6 +343,26 @@ def __init__(
...
+            and self.parallel_config.tensor_parallel_size == 1
+            and self.parallel_config.pipeline_parallel_size == 1
+            and self.parallel_config.prefill_context_parallel_size == 1
+            and self.dcp_size == 1
+            and self.num_ubatches == 2
```
Comment (specture724) [inline vllm/v1/worker/gpu/ubatch_utils.py]: I wonder if we can make it less strict, not only for 2 ubatches?

**Review by specture724 — COMMENTED** [review]
Thank @algebra-MCX for the optimization, which leads to a significant throughput increase.
To conclude my review, the staging can be generalized (odd graph sizes, more than two microbatches)

The `% 2 == 0` and `num_ubatches == 2` restrictions come from hard-coding `capacity = N // 2` and two equal halves. The captured graph is already split by `create_ubatch_slices` (`sp = N // k`, the last region takes the remainder), so staging can reuse those slices:

- Spread the `n` single-token requests evenly: `counts = [n // k + (i < n % k) for i in range(k)]`. Each count is `<= ceil(n / k) <= sp`, so it always fits.
- Stage with one scatter per buffer instead of copy/zero/fill loops. This stays correct if source and destination overlap (possible with `k > 2`):

  ```python
  staged = torch.zeros_like(buf[:num_padded])
  staged[dst] = buf[:n]
  buf[:num_padded].copy_(staged)
  ```

  Slot mappings do the same with `-1` fill, and `is_padding[dst] = False`.
- Restoring the output order becomes one line, replacing `compact_conditional_output` (whose `capacity: int = 64` default looks like a leftover):

  ```python
  model_output[: staged_rows.numel()] = model_output[staged_rows]
  ```
- The DP gate generalizes to "some rank's last microbatch would be empty":

  ```python
  min_tokens = int(num_tokens_across_dp.min())
  if min_tokens <= ubatch_desc.num_tokens // num_ubatches * (num_ubatches - 1):
      runner = cudagraph_manager.ubatch_runner
      if not (
          ubatch_desc.cg_mode == CUDAGraphMode.FULL
          and runner is not None
          and runner.stage_real_tokens
          and synced_uniform_token_count == 1
          and min_tokens >= num_ubatches
      ):
          ubatch_desc = None
  ```

  For `k = 2` this is exactly the current `2 * min <= N`, including odd `N`.
- In `prepare()`, stage when `input_batch.num_tokens <= ubatch_slices[-1].token_slice.start`, and rebuild each microbatch's `query_start_loc` / `seq_lens` with its own region size instead of a shared `capacity`.

**Comment by algebra-MCX** [comment]
Thank you so much for the detailed review and suggestions! @specture724  I’ve already started working on the generalization, but I’ve run into some difficulties.

I’ll keep working on it and update the PR once I’ve made progress. Really appreciate your help!

**Comment by specture724** [comment]
> I’ve run into some difficulties.

Pls share and discuss with us.
